cmd/lotus-chainwatch/storage.go

```diff
@@ -792,6 +840,115 @@ func (st *storage) updateMinerSectors(minerTips map[types.TipSetKey][]*minerStat
 	return updateTx.Commit()
 }
 
+func (st *storage) storeMarketActorDealStates(marketTips map[types.TipSetKey]*marketStateInfo, tipHeights []tipsetKeyHeight, api api.FullNode) error {
+	tx, err := st.db.Begin()
+	if err != nil {
+		return err
+	}
+	if _, err := tx.Exec(`create temp table mds (like market_deal_states excluding constraints) on commit drop;`); err != nil {
+		return err
+	}
+	stmt, err := tx.Prepare(`copy mds (deal_id, sector_start_epoch, last_update_epoch, slash_epoch, state_root) from STDIN`)
+	if err != nil {
+		return err
+	}
+	for _, th := range tipHeights {
+		mt := marketTips[th.tsKey]
+		log.Infow("store deal state", "height", th.height, "tipset", th.tsKey, "minerTS", mt.tsKey)
+		dealStates, err := api.StateMarketDeals(context.TODO(), mt.tsKey)
```

Comment:
Can we move this into syncHead and leave storage primarily for persistence concerns? I think it would be better for us to keep API access outside of storage.go unless there's a good reason for it. It will make it easier to separate these later when we refactor.

Reply:
This call exists here, and below (and in storeMinerSectors) to avoid holding all of the deal state in memory (where # of deal states ~= batch-size) before persisting to the database. Accessing the api here allows for only the deal state at tipset `mt.tsKey` to held in memory for each iteration.

Reply:
I think the batch-size is a nice dial to allow the user to manage chainwatch memory consumption... with batching configurable, I don't mind the memory cost. We don't need to reconcile it now, but when we refactor the code I believe it will probably need to be addressed then. I happy to let you be the judge on whether this warrants attention now or not. 🤝